docs/APP-SNAPSHOT.md

```diff
@@ -0,0 +1,1053 @@
+of the snapshots is consistent with EVE's internal state.
+
+EVE manages snapshots for ext4 file systems by coordinating snapshot operations
+with the application lifecycle and performing low-level snapshot operations
+using the qemu-img tool. Snapshot creation and rollback require stopping and
+reactivating the application to ensure data consistency. Snapshot deletion is
+deferred until the next application reboot but is immediately reported to the
+controller to maintain consistency in the system's view of the snapshots. This
+approach allows EVE to efficiently manage snapshots while minimizing disruptions
+to the application's operation.
+
+### Copy-on-Write (CoW) Approach
+
+The qemu-img tool uses a Copy-on-Write (CoW) approach for snapshot management.
+When a new snapshot is created, it does not immediately occupy a large amount of
+storage. Instead, the snapshot files grow over time as changes are made to the
+file system. The CoW approach ensures that only the differences (or deltas)
+between the current file system state and the snapshot state are stored. As more
+changes are made to the file system, the delta files increase in size. This
+approach allows for efficient storage utilization and minimizes the immediate
+impact on storage capacity when creating new snapshots.
+
+## Handling Purgeable Volumes in EVE
+
+Purgeable volumes in EVE are unique in that they can be entirely replaced with
+new versions during the execution of a purge command. This behavior requires a
+different approach to snapshot management compared to non-purgeable volumes.
+
+### Snapshot Creation for Purgeable Volumes
+
+When a snapshot is created for a purgeable volume, EVE retains the entire old
+volume as a backup. This backup includes all the data in the volume at the time
+of the snapshot. Unlike non-purgeable volumes, where a delta file is typically
+created, purgeable volumes require the entire volume to be preserved. During the
+snapshot creation, EVE retains the original volume file in its existing
+location, ensuring that the old volume is preserved in its entirety.
+
+### Rollback Operation for Purgable Volumes
+
+In the event of a rollback to a previous snapshot, EVE restores the purgeable
+volume by using the original volume file that was retained during the snapshot
+creation. The process involves updating the volume ID in the domain config to
+point to the original volume file, effectively restoring the volume to its state
+at the time of the snapshot.
+
+### Snapshot Deletion for Purgable Volumes
+
+When a snapshot of a purgeable volume is deleted, EVE removes any references to
+the original volume file associated with the snapshot ID.
+
+## Serialization and Deserialization
+
+The snapshotting feature relies on robust serialization and deserialization
+functionality to ensure the continuity of snapshot lists after the system
+reboots. The serialization process converts snapshot data structures into a
```

Comment:
Makes sense to add that this is using persistent storage (I was looking for the string "persist.*' and didn't find this text)

Reply:
Added more details in this section, including the mention of the persist directory.